protocols/imap/src/main/java/org/apache/james/imap/processor/base/UidMsnMapper.java

```diff
@@ -75,7 +79,41 @@ public synchronized void clear() {
     }
 
     public synchronized void addUid(MessageUid uid) {
-        this.addMapping(nextMsn(), uid);
+        if (isLastUid(uid)) {
+            addMapping(nextMsn(), uid);
+        } else {
+            addUidInMiddle(uid);
+        }
+    }
+
+    private boolean isLastUid(MessageUid uid) {
+        return msnToUid.isEmpty() ||
+            uid.asLong() > msnToUid.get(getLastMsn()).asLong();
+    }
+
+    private boolean alreadyContains(MessageUid uid) {
+        return msnToUid.containsValue(uid);
+    }
+
+    private void addUidInMiddle(MessageUid uid) {
+        List<MessageUid> aboveUids = removeAndGetAboveUidSortedInIncreasingOrder(uid);
+        addMapping(nextMsn(), uid);
+        for (MessageUid aboveUid : aboveUids) {
+            addMapping(nextMsn(), aboveUid);
+        }
+    }
+
+    private List<MessageUid> removeAndGetAboveUidSortedInIncreasingOrder(MessageUid uid) {
+        ImmutableList.Builder<MessageUid> result = ImmutableList.builder();
+        int position = getLastMsn();
+        Optional<MessageUid> maxUid = getUid(position);
+        while (maxUid.isPresent() && uid.asLong() < maxUid.get().asLong()) {
```

Comment:
I had a look at priority queue, maybe expressed in a simpler way but we can not afford O(n) computation on every read for very large mailboxes. Thus I don't consider it as an option.

What we would like apparently is some sorted list (to have directly access to the item position... ). There seems no JDK collection matches. I'll have a look in Guava.